web_server.go

```diff
@@ -42,14 +47,31 @@ func newWebServer(
 	auth *Auth,
 	keychain *keychain.Keychain,
 	maxRequestSize int64,
-	www string,
+	baseURL string,
+	webDir string,
 	header http.Header,
-) *WebServer {
-	fs := handleStatic("/", http.FileServer(http.Dir(www)), header)
+) (*WebServer, error) {
+
+	// read default index.html page from the web root
+	indexPage, err := ioutil.ReadFile(filepath.Join(webDir, "index.html"))
+	if err != nil {
+		return nil, fmt.Errorf("failed reading default index.html page: %v", err)
+	}
+
+	fs := handleStatic([]byte(mungeIndexPage(baseURL, string(indexPage))), http.StripPrefix(baseURL, http.FileServer(http.Dir(webDir))), header)
 	if auth != nil {
 		fs = auth.wrap(fs)
 	}
-	return &WebServer{site, broker, fs, keychain, maxRequestSize}
+	return &WebServer{site, broker, fs, keychain, maxRequestSize, baseURL}, nil
+}
+
+func mungeIndexPage(baseURL, html string) string {
```

Comment:
My original intention was to use relative frontend paths and just rewrite / strip path prefixes server side. So for example ws connection (same for file API endpoint, static resources) would connect to `_s` instead of absolute `/_s`. Assuming browser url would be `http:localhost:10101/prefix/app-route`, the final ws connection string would be `http:localhost:10101/prefix/app-route/_s`. Server would be then able to strip the prefix as it is aware of `base-url` and `app-route` and correctly route it to `/_s` handler.

This would save us from munging `index.html` and keep frontend part simple.

wdyt @lo5 ?

Reply:
The go routing mechanism expects rooted paths, though. So a page accessed at /a/b/c would attempt to connect to /a/b/c/_s, which will not hit the websocket handler.

Reply:
But your strategy could be made to work if we bypass go's routing altogether and implement a custom routing mechanism.